src/network/networkctl.c

```diff
@@ -1395,87 +1395,71 @@ static int dump_dhcp_leases(Table *table, const char *prefix, sd_bus *bus, const
-                               r, "Failed to query link DHCP leases: %s", bus_error_message(&error, r));
-                return 0;
-        }
-
-        r = sd_bus_message_enter_container(reply, 'v', "a(uayayayayt)");
+        r = bus_call_method(bus, bus_network_mgr, "DescribeLink", &error, &reply, "i", link->ifindex);
         if (r < 0)
-                return bus_log_parse_error(r);
+                return log_error_errno(r, "Failed to get link description: %s", bus_error_message(&error, r));
 
-        r = sd_bus_message_enter_container(reply, 'a', "(uayayayayt)");
+        r = sd_bus_message_read(reply, "s", &text);
         if (r < 0)
                 return bus_log_parse_error(r);
 
-        while ((r = sd_bus_message_enter_container(reply, 'r', "uayayayayt")) > 0) {
-                _cleanup_free_ char *id = NULL, *ip = NULL;
-                const void *client_id, *addr, *gtw, *hwaddr;
-                size_t client_id_sz, sz;
-                uint64_t expiration;
-                uint32_t family;
-
-                r = sd_bus_message_read(reply, "u", &family);
-                if (r < 0)
-                        return bus_log_parse_error(r);
-
-                r = sd_bus_message_read_array(reply, 'y', &client_id, &client_id_sz);
-                if (r < 0)
-                        return bus_log_parse_error(r);
-
-                r = sd_bus_message_read_array(reply, 'y', &addr, &sz);
-                if (r < 0 || sz != 4)
-                        return bus_log_parse_error(r);
-
-                r = sd_bus_message_read_array(reply, 'y', &gtw, &sz);
-                if (r < 0 || sz != 4)
-                        return bus_log_parse_error(r);
-
-                r = sd_bus_message_read_array(reply, 'y', &hwaddr, &sz);
-                if (r < 0)
-                        return bus_log_parse_error(r);
+        r = json_parse(text, 0, &interface, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse JSON: %m");
 
-                r = sd_bus_message_read_basic(reply, 't', &expiration);
-                if (r < 0)
-                        return bus_log_parse_error(r);
+        dhcp_server = json_variant_by_key(interface, "DHCPServer");
+        if (dhcp_server) {
+                JSON_VARIANT_ARRAY_FOREACH(lease, json_variant_by_key(dhcp_server, "Leases")) {
+                        JsonVariant *hostname = NULL;
+                        _cleanup_free_ uint8_t *client_id = NULL, *addr = NULL;
+                        _cleanup_free_ char *id = NULL, *ip = NULL;
+                        size_t client_id_sz, addr_sz;
+                        uint64_t expiration_usec;
+
+                        r = json_variant_bytes(
+                                        json_variant_by_key(lease, "ClientId"), &client_id, &client_id_sz);
+                        if (r < 0)
+                                return r;
+                        r = sd_dhcp_client_id_to_string((void *) client_id, client_id_sz, &id);
```

Comment:
This will fail with assertion when `json_variant_bytes()` returns zero length client ID, that is when the json does not have ClientId field.
Maybe, better to skip the lease or return -EINVAL when `!client_id || client_id_sz == 0`.